pkg/cluster/deploystorage.go

```diff
@@ -148,6 +157,19 @@ func (m *manager) deployStorageTemplate(ctx context.Context) error {
 	return arm.DeployTemplate(ctx, m.log, m.deployments, resourceGroup, "storage", t, nil)
 }
 
+func (m *manager) imageRegistryStorageAccountName() (string, error) {
+	if m == nil {
+		return "", fmt.Errorf("cluster manager is nil")
+	}
```

Comment:
You can reach that point as you can not ensure that m was initialized before invoking the method.
Here is an example with a nil pointer exception: 
```go
package main

import (
	"log"
)

type manager struct {
}

func (m *manager) Foo() {
	log.Print(m)
	log.Print(*m) // <- this line will panic if pointer receiver (m) was NOT initialised
	log.Print(&m)
	log.Print(&(*m))
}

func main() {
	var m *manager = &manager{}
	m.Foo() // <- no panic as the pointer receiver is initialised

	log.Print("\n\n")

	var m2 *manager
	m2.Foo()
}

```
And here is that example in the [Playground](https://go.dev/play/p/dD3KdnFpIqU)

The line that is giving nil is
```go
log.Print(*m)
```
because you try to dereference a non-nil pointer.

If you add 
```go
if m == nil {
    return
}
```
inside Foo() in the example provided it will not panic.
🙂

Reply:
but any references to that function has at least one instance of accessing fields referenced by that pointer. it will never reach that point if it is nil. 
Here nil is not a valid value for the manager, so you should check if it is nil at creation if you do. it would become incredibly cumbersome to check it for every pointer receiver. And by not valid, I mean that we can't do anything of it, not that we can't go out of our way to create a nil pointer.

Reply:
"it will never reach that point if it is nil." ->  You can call m.imageRegistryStorageAccountName() when m (of type *manager) have been declared but not initialised, so it is possible to reach that point. 

We can not assume that m will never be nil when calling m.imageRegistryStorageAccountName(), so you must check if m == nil inside the function or outside, but you have to check it at some point. 

The good thing about checking for nil inside the imageRegistryStorageAccountName() is that you would never need to check if m is nil before calling this function. So if in your codebase you have N calls to m.imageRegistryStorageAccountName() you just need to do the nil check inside the function, not N times before calling the function.

I am removing the nil check inside the function because it is something we don't usually do in the codebase but still think it has some interesting usages. 

🙂

Reply:
if there is a check it should be at creation time, not inside the function since we are never expecting this to be a nil pointer. Here it makes no sense to call that function on a nil pointer, if it is nil, something failed before and we should have failed beforehand

Reply:
"Here it makes no sense to call that function on a nil pointer" ->In fact, it almost never makes sense to call a method on a pointer receiver so it is not specific to that case. As soon as you dereference the pointer inside the function, you should ensure it is not nil.

Obviously, it is better to assert correctness as close to the creation of entities as possible, that's for sure. I completely get your point about checking things before we reach a point where we call to the method on a nil receiver, BUT...

The problem is that even if you do ALL that, nothing prevents the codebase to have in the package cluster something like:

```go
var m *manager
// more code...
m. imageRegistryStorageAccountName()
```
and panic will occur. 

Because m. imageRegistryStorageAccountName() can be called anywhere in the cluster package, you need to check for not nil before calling the method, doing it inside or outside the function is your decision, but is a decision that you should make. 

But again, I removed the check inside the function so we are all happy 🙂

Reply:
Sorry, when I said "In fact, it almost never makes sense to call a method on a pointer receiver so it is not specific to that case."
I wanted to say
"In fact, it almost never makes sense to call a method on a NIL pointer receiver so it is not specific to that case." (I forgot the nil), as I wanted to respond to your sentence "Here it makes no sense to call that function on a nil pointer".

Nothing against pointer receivers, just against never checking for nil receivers.

All clear.